File: Chess_Robot/src/chess_robot/Chessboard.java
package chess_robot;

import java.util.Scanner;
import java.util.LinkedList;

import stockfish.Stockfish;

class chessPos {

	int x, y;

	chessPos() {
		x = 0;
		y = 0;
	}

	chessPos(int x, int y) {
		this.x = x;
		this.y = y;
	}

	chessPos(chessPos pos) {
		set(pos);
	}

	void set(chessPos pos) {
		x = pos.x;
		y = pos.y;
	}

	// 1, 3, ... , 17: gap / 0, 2, ... , 18: piece
	boolean valid() {
		return 0 <= x && x <= 18 && 0 <= y && y <= 18;
	}

	// a1 -> 16, 2
	void set(String name) {
		x = (9 - Character.digit(name.charAt(1), 10)) * 2;
		y = (Character.digit(name.charAt(0), 18) - 9) * 2;
	}

	String name() {
		if (x % 2 != 0 || y % 2 != 0 || x == 0 || x == 18 || y == 0 || y == 18)
			return "";
		return String.format("%c%c", Character.forDigit((y / 2 + 9), 18),
				Character.forDigit((9 - x / 2), 10));
	}

	void input(Scanner sc, String msg) {
		System.out.print(msg);
		try {
			String name = sc.nextLine();
			if (name.length() != 2)
				throw new Exception();
			set(name);
			if (valid())
				return;
		} catch (Exception e) {
		}
		System.out.println("wrong input; wrong format");
		input(sc, msg);
	}

	void print() {
		System.out.printf("(%d, %d)", x, y);
	}

	void move(dir dir, boolean forward) {
		x += dir.dx * (forward ? 1 : -1);
		y += dir.dy * (forward ? 1 : -1);
	}

	boolean isSame(chessPos pos) {
		return x == pos.x && y == pos.y;
	}
}

enum dir {
	up(-1, 0), right(0, 1), down(1, 0), left(0, -1);

	int dx, dy;

	dir(int dx, int dy) {
		this.dx = dx;
		this.dy = dy;
	};
};

public class Chessboard {

	// lower case: black, upper case: white
	// King, Queen, Rook, Bishop, kNight, Pawn
	static final String PIECE = " rnbqkpRNBQKP";

	// 4 = extra space for piece captured
	static char[][] board = new char[15 + 4][15 + 4];
	static String fen = "";
	static boolean isUserTurn, isOver = false;

	static void initBoard() {
		for (int i = 0; i < 19; i++)
			for (int j = 0; j < 19; j++)
				board[i][j] = ' ';
		board[2] = "  r n b q k b n r  ".toCharArray();
		board[4] = "  p p p p p p p p  ".toCharArray();
		board[14] = "  P P P P P P P P  ".toCharArray();
		board[16] = "  R N B Q K B N R  ".toCharArray();
	}

	/*
	 * static String fen() { String fen = ""; int count; for (int i = 2; i < 17;
	 * i += 2) { count = 0; for (int j = 2; j < 17; j += 2) { if (board[i][j] ==
	 * ' ') count++; else { if (count != 0) { fen += count; count = 0; } fen +=
	 * board[i][j]; } } if (count != 0) fen += count; if (i != 16) fen += '/'; }
	 * return fen; }
	 */

	static void printBoard() {
		System.out.println("\n      a b c d e f g h");
		System.out.println("      ---------------");
		System.out.println("    " + String.valueOf(board[0]));
		for (int i = 2; i < 17; i += 2) {
			System.out.printf("%d | ", 9 - i / 2);
			System.out.println(board[i]);
		}
		System.out.println("    " + String.valueOf(board[18]));
		System.out.println("Fen: " + fen + "\n");
	}

	static boolean canGo(chessPos pos) {
		if (!pos.valid())
			return false;
		if (board[pos.x][pos.y] != ' ')
			return false;
		return true;
	}

	static boolean isWhiteTurn() {
		return fen.contains("w");
	}

	static chessPos findBlank() {
		int i;
		if (isWhiteTurn()) {
			for (i = 0; i < 19; i += 2)
				if (board[0][i] == ' ')
					return new chessPos(0, i);
			for (i = 2; i < 9; i += 2) {
				if (board[i][0] == ' ')
					return new chessPos(i, 0);
				if (board[i][18] == ' ')
					return new chessPos(i, 18);
			}
		} else {
			for (i = 0; i < 19; i += 2)
				if (board[18][i] == ' ')
					return new chessPos(18, i);
			for (i = 16; i > 9; i -= 2) {
				if (board[i][0] == ' ')
					return new chessPos(i, 0);
				if (board[i][18] == ' ')
					return new chessPos(i, 18);
			}
		}
		// no case for this
		return new chessPos();
	}

	static void movePiece(chessPos startPos, chessPos endPos) {
		if (!canGo(endPos))
			movePiece(endPos, findBlank());

		boolean visited[][] = new boolean[15 + 4][15 + 4];
		dir direction[][] = new dir[15 + 4][15 + 4];
		LinkedList<chessPos> queue = new LinkedList<chessPos>();
		LinkedList<dir> path = new LinkedList<dir>();
		chessPos pos = new chessPos(startPos);

		visited[pos.x][pos.y] = true;
		queue.add(pos);

		while (queue.size() != 0) {
			pos = queue.poll();

			for (int i = 0; i < 4; i++) {
				chessPos nextPos = new chessPos(pos);
				dir nextDir = dir.values()[i];

				nextPos.move(nextDir, true);
				if (canGo(nextPos) && !visited[nextPos.x][nextPos.y]) {
					visited[nextPos.x][nextPos.y] = true;
					direction[nextPos.x][nextPos.y] = nextDir;
					queue.add(nextPos);
				}
			}
		}

		pos.set(endPos);
		while (!pos.isSame(startPos)) {
			path.addFirst(direction[pos.x][pos.y]);
			pos.move(direction[pos.x][pos.y], false);
		}

		System.out.print("path for " + startPos.name() + " -> " + endPos.name()
				+ ": ");
		pos.set(startPos);
		for (dir dir : path) {
			pos.print();
			System.out.print(" -> " + dir + " -> ");
			pos.move(dir, true);
		}
		pos.print();
		System.out.println();

		board[endPos.x][endPos.y] = board[startPos.x][startPos.y];
		board[startPos.x][startPos.y] = ' ';
	}

	static boolean isMoveCompleted() {
		return true;
	}

	static boolean setTurn(Scanner sc) {
		System.out.println("who will go first; user(0) / AI(1)");
		try {
			String turn = sc.nextLine();
			if (turn.equals("0")) {
				return true;
			} else if (turn.equals("1")) {
				return false;
			}
		} catch (Exception e) {
		}
		System.out.println("wrong input; choose between 0 and 1");
		return setTurn(sc);
	}

	public static void main(String[] args) {
		initBoard();
		printBoard();

		Stockfish client = new Stockfish();
		if (client.startEngine()) {
			System.out.println("start Engine");

			client.sendCommand("uci");
			client.getOutput(0, false);

			client.sendCommand("position startPos");
			fen = client.getFen();

			Scanner sc = new Scanner(System.in);

			isUserTurn = setTurn(sc);

			while (!isOver) {
				chessPos startPos = new chessPos(), endPos = new chessPos();
				String move, nextFen;

				if (isUserTurn) {
					startPos.input(sc, "from: ");
					endPos.input(sc, "to: ");

					// quit game
					if (startPos.isSame(endPos))
						break;

					move = startPos.name() + endPos.name();
					nextFen = client.moveAndGetFen(fen, move);

					if (fen.equals(nextFen)) {
						System.out.println("wrong input; can't move");
					} else {
						fen = nextFen;
						isUserTurn = false;
						movePiece(startPos, endPos);
					}

				} else {
					move = client.getBestMove(fen, 100);

					fen = client.moveAndGetFen(fen, move);
					isUserTurn = true;

					startPos.set(move.substring(0, 2));
					endPos.set(move.substring(2));

					movePiece(startPos, endPos);
				}

				printBoard();
				
				move = client.getBestMove(fen, 100);

				if (move.equals("(none)")) {
					isOver = true;
					System.out.println("checkmate; game over");
				} 
			}

			client.stopEngine();
		} else
			System.out.println("can't start Engine");
	}
}

File: chessRobot/chessboard.py
from enum import Enum
from stockfish import Stockfish

class ChessPos:
    def __init__(self, x=0, y=0):
        self.x = x
        self.y = y
    
    def set(self, pos):
        self.x = pos.x
        self.y = pos.y
    
    # 1, 3, ... , 17: gap / 0, 2, ... , 18: piece
    def valid(self):
        return 0 <= self.x <= 18 and 0 <= self.y <= 18
    
    # a1 -> 16, 2
    def setFromName(self, name):
        self.x = (9 - int(name[1])) * 2
        self.y = (ord(name[0]) - ord('a') + 1) * 2
    
    def name(self):
        if self.x % 2 != 0 or self.y % 2 != 0 or self.x == 0 or self.x == 18 or self.y == 0 or self.y == 18:
            return ""
        return f"{chr(self.y // 2 - 1 + ord('a'))}{int(9 - self.x // 2)}"
    
    def input(self, msg):
        try:
            name = input(msg)
            if len(name) != 2:
                raise Exception
            self.setFromName(name)
            if self.valid():
                return
        except:
            pass
        self.input(msg)
    
    def print(self):
        print(f"({self.x}, {self.y})", end='')
    
    def move(self, direction, forward):
        self.x += direction.value[0] * (1 if forward else -1)
        self.y += direction.value[1] * (1 if forward else -1)
    
    def isSame(self, pos):
        return self.x == pos.x and self.y == pos.y

class Direction(Enum):
    up = (-1, 0)
    right = (0, 1)
    down = (1, 0)
    left = (0, -1)

class Chessboard:

    # lower case: black, upper case: white
	# King, Queen, Rook, Bishop, kNight, Pawn
    PIECE = " rnbqkpRNBQKP"

    # 4 = extra space for piece captured
    board = [[' '] * (15 + 4) for _ in range(15 + 4)]
    prevFen = ""
    fen = ""
    extra = ""
    isUserTurn = True
    isOver = False
    startPos = ChessPos()
    endPos = ChessPos()
    robotPos = ChessPos()

    @classmethod
    def toMove(cls):
        return cls.startPos.name() + cls.endPos.name() + cls.extra

    @classmethod
    def fromMove(cls, move):
        cls.startPos.setFromName(move[:2])
        cls.endPos.setFromName(move[2:4])
        cls.extra = move[4:]
    
    @classmethod
    def initBoard(cls):
        cls.board[2] = list("  r n b q k b n r  ")
        cls.board[4] = list("  p p p p p p p p  ")
        cls.board[14] = list("  P P P P P P P P  ")
        cls.board[16] = list("  R N B Q K B N R  ")

    @classmethod
    def printBoard(cls):
        print("\n      a b c d e f g h")
        print("      ---------------")
        print("    " + ''.join(cls.board[0]))
        for i in range(2, 17, 2):
            print(f"{9 - i // 2} | " + ''.join(cls.board[i]))
        print("    " + ''.join(cls.board[18]))
        print("Fen: " + cls.fen + "\n")

    @classmethod
    def canGo(cls, pos, moveRobot = False):
        if not pos.valid():
            return False
        if not moveRobot and cls.board[pos.x][pos.y] != ' ':
            return False
        return True

    @classmethod
    def isWhiteTurn(cls):
        return 'w' in cls.fen

    @classmethod
    def findBlank(cls):
        if cls.isWhiteTurn():
            '''for i in range(2, 17, 2):
                if cls.board[0][i] == ' ':
                    return ChessPos(0, i)'''
            for i in range(2, 9, 2):
                if cls.board[i][0] == ' ':
                    return ChessPos(i, 0)
                if cls.board[i][18] == ' ':
                    return ChessPos(i, 18)
        else:
            '''for i in range(2, 17, 2):
                if cls.board[18][i] == ' ':
                    return ChessPos(18, i)'''
            for i in range(16, 9, -2):
                if cls.board[i][0] == ' ':
                    return ChessPos(i, 0)
                if cls.board[i][18] == ' ':
                    return ChessPos(i, 18)
        # no case for this
        return ChessPos()

    @classmethod
    def movePiece(cls, startPos, endPos, moveRobot = False):
        visited = [[False] * (15 + 4) for _ in range(15 + 4)]
        direction = [[None] * (15 + 4) for _ in range(15 + 4)]
        queue = []
        path = []
        pos = ChessPos(startPos.x, startPos.y)

        visited[pos.x][pos.y] = True
        queue.append(pos)

        while queue:
            pos = queue.pop(0)

            for dir in Direction:
                nextPos = ChessPos(pos.x, pos.y)
                nextDir = dir

                nextPos.move(nextDir, True)
                if cls.canGo(nextPos, moveRobot) and not visited[nextPos.x][nextPos.y]:
                    visited[nextPos.x][nextPos.y] = True
                    direction[nextPos.x][nextPos.y] = nextDir
                    queue.append(nextPos)

        pos.set(endPos)
        while not pos.isSame(startPos):
            path.insert(0, direction[pos.x][pos.y])
            pos.move(direction[pos.x][pos.y], False)

        print(moveRobot)
        print(f"path for {startPos.name()} -> {endPos.name()}: ", end='')
        pos.set(startPos)
        for dir in path:
            pos.print()
            print(f" -> {dir.name} -> ", end='')
            pos.move(dir, True)
        pos.print()
        print()

        if not moveRobot:
            cls.board[endPos.x][endPos.y] = cls.board[startPos.x][startPos.y]
            cls.board[startPos.x][startPos.y] = ' '
        cls.robotPos.set(endPos)

    @classmethod
    def go(cls):
        cls.movePiece(cls.robotPos, cls.startPos, True)

        if not cls.canGo(cls.endPos):
            cls.movePiece(cls.endPos, cls.findBlank())

        if cls.extra:
            cls.board[cls.startPos.x][cls.startPos.y] = cls.extra[0]

        cls.movePiece(cls.startPos, cls.endPos)

        if cls.board[cls.endPos.x][cls.endPos.y] in 'kK':
            if cls.endPos.y - cls.startPos.y == 4:
                cls.movePiece(ChessPos(cls.endPos.x, 16), ChessPos(cls.endPos.x, 12))
            if cls.startPos.y - cls.endPos.y == 4:
                cls.movePiece(ChessPos(cls.endPos.x, 2), ChessPos(cls.endPos.x, 8))

        if cls.prevFen.split(" ")[3] != "-" and cls.fen.split(" ")[3] == "-":
            if cls.board[cls.endPos.x][cls.endPos.y] == 'p':
                cls.movePiece(ChessPos(cls.endPos.x - 2, cls.endPos.y), cls.findBlank())
            if cls.board[cls.endPos.x][cls.endPos.y] == 'P':
                cls.movePiece(ChessPos(cls.endPos.x + 2, cls.endPos.y), cls.findBlank())

    @classmethod
    def isMoveCompleted(cls):
        return True

    @classmethod
    def setTurn(cls):
        try:
            turn = int(input("who will go first; user(0) / AI(1)\n"))
            if turn == 0:
                return True
            elif turn == 1:
                return False
        except Exception as e:
            print(e)
        print("wrong input; choose between 0 and 1")
        return cls.setTurn()

    @classmethod
    def inputMove(cls):
        try:
            pos = input("move: ")
            cls.startPos.setFromName(pos[:2])
            cls.endPos.setFromName(pos[2:4])
            cls.extra = pos[4:]
            if cls.startPos.valid() and cls.endPos.valid():
                return cls.toMove()
        except Exception as e:
            print(e)
        print("wrong input; wrong format")
        return cls.inputMove()

    @classmethod
    def quit(cls):
        return cls.startPos.isSame(cls.endPos)

    @classmethod
    def main(cls):
        cls.initBoard()
        cls.printBoard()

        client = Stockfish()
        if client.startEngine():
            print("start engine")

            client.sendCommand("uci")
            client.getOutput(0, False)

            client.sendCommand("position startPos")
            cls.fen = client.getFen()

            cls.isUserTurn = cls.setTurn()

            while not cls.isOver:
                move, nextFen = "", ""

                if cls.isUserTurn:
                    move = cls.inputMove()
                else:
                    move = client.getBestMove(100)
                    cls.fromMove(move)

                if cls.quit():
                    break

                nextFen = client.moveAndGetFen(cls.fen, move)

                if cls.fen == nextFen:
                    print("wrong input; can't move as above")
                else:
                    cls.prevFen = cls.fen
                    cls.fen = nextFen
                    cls.isUserTurn = not cls.isUserTurn

                    cls.go()

                cls.printBoard()

                move = client.getBestMove(100)

                if move == "(none)":
                    cls.isOver = True
                    print("game over")

            client.stopEngine()
            print("stop engine")
        else:
            print("can't start engine")

if __name__ == "__main__":
    Chessboard.main()
